utils/benchmarking/eventing/timeseriesdb/main.go

```diff
@@ -0,0 +1,151 @@
+// MIT License
+//
+// Copyright (c) 2021 Mert Bora Alper and EASE lab
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+package main
+
+import (
+	"context"
+	"log"
+	"net"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
+	"google.golang.org/protobuf/types/known/durationpb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"eventing/proto"
+)
+
+type Server struct {
+	proto.UnimplementedTimeseriesServer
+
+	completionEventDescriptorsT []*proto.CompletionEventDescriptor
+	completionEventDescriptors  map[string][]*proto.CompletionEventDescriptor
+	invocations                 map[string]*proto.InvocationDescriptor
+}
+
+func (s *Server) StartExperiment(_ context.Context, definition *proto.ExperimentDefinition) (*empty.Empty, error) {
+	log.Printf("starting experiment %+v\n", definition.CompletionEventDescriptors)
+
+	s.completionEventDescriptorsT = make([]*proto.CompletionEventDescriptor, len(definition.CompletionEventDescriptors))
+	copy(s.completionEventDescriptorsT, definition.CompletionEventDescriptors)
+	s.completionEventDescriptors = make(map[string][]*proto.CompletionEventDescriptor)
+	s.invocations = make(map[string]*proto.InvocationDescriptor)
+
+	return &empty.Empty{}, nil
+}
+
+func (s *Server) EndExperiment(_ context.Context, _ *empty.Empty) (*proto.ExperimentResult, error) {
+	log.Println("ending current experiment")
+	invocations := make([]*proto.InvocationDescriptor, 0)
+	for _, inv := range s.invocations {
+		invocations = append(invocations, inv)
+	}
+	return &proto.ExperimentResult{Invocations: invocations}, nil
+}
+
+func (s *Server) RegisterEvent(_ context.Context, event *proto.Event) (*empty.Empty, error) {
```

Comment:
this function is too long, create helper functions. If a function doesn't fit one screen, it should be refactored